src/core/state.go

```diff
@@ -268,135 +232,94 @@ func (state *BuildState) addActiveTargets(n int) {
-// We retain the internal priority queue since it is unbounded size which is pretty important
-// for us not to deadlock.
-func (state *BuildState) feedQueues(parses ParseTaskQueue, builds BuildTaskQueue, tests TestTaskQueue, remoteBuilds BuildTaskQueue, remoteTests TestTaskQueue) {
-	anyRemote := state.Config.NumRemoteExecutors() > 0
-	for {
-		t, _ := state.pendingTasks.Get(1)
-		task := t[0].(pendingTask)
-		remote := func() bool {
-			return anyRemote && !state.Graph.Target(task.Label).Local
+func (state *BuildState) addPendingTest(target *BuildTarget, numRuns int) {
+	atomic.AddInt64(&state.progress.numPending, int64(numRuns))
+	go func() {
+		defer func() {
+			recover() // Prevent death on 'send on closed channel'
+		}()
+		ch := state.pendingTests
+		if state.anyRemote && !target.Local {
+			ch = state.pendingRemoteTests
 		}
-
-		switch task.Type {
-		case Stop, Kill:
-			close(parses)
-			close(builds)
-			close(tests)
-			close(remoteBuilds)
-			close(remoteTests)
-			return
-		case Parse, SubincludeParse:
-			parses <- ParseTask{Label: task.Label, Dependent: task.Dependent, ForSubinclude: task.Type == SubincludeParse}
-		case Build, SubincludeBuild:
-			atomic.AddInt64(&state.progress.numRunning, 1)
-			if remote() {
-				remoteBuilds <- task.Label
-			} else {
-				builds <- task.Label
-			}
-		case Test:
-			atomic.AddInt64(&state.progress.numRunning, 1)
-			testTask := TestTask{
-				Label: task.Label,
-				Run:   task.Run,
-			}
-			if remote() {
-				remoteTests <- testTask
-			} else {
-				tests <- testTask
-			}
+		for run := 1; run <= numRuns; run++ {
+			ch <- TestTask{Label: target.Label, Run: run}
 		}
-	}
-}
-
-func (state *BuildState) addPending(label BuildLabel, t taskType) {
-	state.addPendingTask(pendingTask{Label: label, Type: t})
+	}()
 }
 
-func (state *BuildState) addPendingTask(task pendingTask) {
-	atomic.AddInt64(&state.progress.numPending, 1)
-	_ = state.pendingTasks.Put(task)
+// TaskQueues returns a set of channels to listen on for tasks of various types.
+func (state *BuildState) TaskQueues() (parses <-chan ParseTask, builds, remoteBuilds <-chan BuildTask, tests, remoteTests <-chan TestTask) {
+	return state.pendingParses, state.pendingBuilds, state.pendingRemoteBuilds, state.pendingTests, state.pendingRemoteTests
 }
 
 // TaskDone indicates that a single task is finished. Should be called after one is finished with
-// a task returned from NextTask(), or from a call to ExtraTask().
-func (state *BuildState) TaskDone(wasBuildOrTest bool) {
-	state.taskDone(false, wasBuildOrTest)
+// a task returned from NextTask().
+func (state *BuildState) TaskDone() {
+	state.taskDone(false)
 }
 
-func (state *BuildState) taskDone(wasSynthetic, wasBuildOrTest bool) {
+func (state *BuildState) taskDone(wasSynthetic bool) {
 	if !wasSynthetic {
 		atomic.AddInt64(&state.progress.numDone, 1)
 	}
-	if wasBuildOrTest {
-		atomic.AddInt64(&state.progress.numRunning, -1)
-	}
```

Comment:
Why do we not decrement this anymore? TaskDone is always called for a build or test now.